source/extensions/filters/http/jwt_authn/jwks_cache.cc

```diff
@@ -98,21 +98,56 @@ class JwksCacheImpl : public JwksCache {
   JwksCacheImpl(const JwtAuthentication& config) {
     for (const auto& it : config.providers()) {
       const auto& provider = it.second;
-      jwks_data_map_.emplace(provider.issuer(), provider);
+      jwks_data_map_.emplace(it.first, provider);
+      if (issuer_ptr_map_.find(provider.issuer()) == issuer_ptr_map_.end()) {
+        issuer_ptr_map_.emplace(provider.issuer(), findByProviderHelper(it.first));
+      }
+    }
+  }
+
+  JwksData* findByIssuer(const std::string& issuer) const override {
+    const auto it = issuer_ptr_map_.find(issuer);
+    if (it == issuer_ptr_map_.end()) {
+      return nullptr;
     }
+    return it->second;
+  }
+
+  JwksData* findByProvider(const std::string& provider) override {
+    return findByProviderHelper(provider);
   }
 
-  JwksData* findByIssuer(const std::string& name) override {
-    auto it = jwks_data_map_.find(name);
+  const AudienceChecker& getAudienceCheckerByProvider(const std::string& provider) const override {
+    const auto verifier = findByProviderHelper(provider);
+    ASSERT(verifier != nullptr);
+    return *verifier;
+  }
+
+  // get provider data by provider name. Returns nullptr if issuer is not found.
+  const AudienceChecker& getAudienceCheckerByIssuer(const std::string& issuer) const override {
+    const auto verifier = findByIssuer(issuer);
+    ASSERT(verifier != nullptr);
+    return *verifier;
+  }
+
+private:
+  JwksDataImpl* findByProviderHelper(const std::string& provider) {
+    return const_cast<JwksDataImpl*>(
+        static_cast<const JwksCacheImpl*>(this)->findByProviderHelper(provider));
```

Comment:
I am not sure if this const_cast will work.    You are casting const T*   to T,  feel not safe.
 
I feel you need to use iterator  version,   not the const_iterator